src/Native/Unix/System.Native/pal_console.c

```diff
@@ -71,92 +73,200 @@ void SystemNative_SetKeypadXmit(const char* terminfoString)
+    termios.c_cc[VTIME] = decisecondsTimeout;
+
+    // Check if the settings have changed.
+    if (g_hasCurrentTermios)
     {
-        if (!g_readInProgress)
+        if (g_currentTermios.c_lflag == termios.c_lflag &&
+            g_currentTermios.c_iflag == termios.c_iflag &&
+            g_currentTermios.c_cc[VMIN] == termios.c_cc[VMIN] &&
+            g_currentTermios.c_cc[VMIN] == termios.c_cc[VMIN])
         {
-            // Store the original settings, but only if we didn't already.  This function
-            // may be called when the process is resumed after being suspended, and if
-            // that happens during a read, we'll call this function to reset the attrs.
-            g_preReadTermios = newTermios;
+            return true;
         }
+    }
+
+    return TcSetAttr(&termios, blockIfBackground);
+}
 
-        newTermios.c_iflag &= (uint32_t)(~(IXON | IXOFF));
-        newTermios.c_lflag &= (uint32_t)(~(ECHO | ICANON | IEXTEN));
-        newTermios.c_cc[VMIN] = minChars;
-        newTermios.c_cc[VTIME] = decisecondsTimeout;
-        IncorporateBreak(&newTermios, g_signalForBreak);
+void UninitializeTerminal()
+{
+    assert(g_haveInitTermios);
+
+    // This method is called on SIGQUIT/SIGINT from the signal dispatching thread
+    // and on atexit.
 
-        if (tcsetattr(STDIN_FILENO, TCSANOW, &newTermios) >= 0)
+    if (pthread_mutex_lock(&g_lock) == 0)
+    {
+        if (!g_terminalUninitialized)
         {
-            g_currTermios = newTermios;
-            g_readInProgress = true;
+            TcSetAttr(&g_initTermios, /* blockIfBackground */ false);
+
+            g_terminalUninitialized = true;
         }
+
+        pthread_mutex_unlock(&g_lock);
+    }
+}
+
+void SystemNative_InitializeConsoleBeforeRead(uint8_t minChars, uint8_t decisecondsTimeout)
+{
+    if (pthread_mutex_lock(&g_lock) == 0)
+    {
+        g_reading = true;
+
+        ConfigureTerminal(g_signalForBreak, /* forChild */ false, minChars, decisecondsTimeout, /* blockIfBackground */ true);
+
+        pthread_mutex_unlock(&g_lock);
     }
 }
 
 void SystemNative_UninitializeConsoleAfterRead()
 {
-    if (g_readInProgress)
+    if (pthread_mutex_lock(&g_lock) == 0)
     {
-        g_readInProgress = false;
-
-        int tmpErrno = errno; // preserve any errors from before uninitializing
-        IncorporateBreak(&g_preReadTermios, g_signalForBreak);
-        int ret = tcsetattr(STDIN_FILENO, TCSANOW, &g_preReadTermios);
-        assert(ret >= 0); // shouldn't fail, but if it does we don't want to fail in release
-        (void)ret;
-        errno = tmpErrno;
+        g_reading = false;
+
+        pthread_mutex_unlock(&g_lock);
+    }
+}
+
+void SystemNative_ConfigureTerminalForChildProcess(int32_t childUsesTerminal)
+{
+    assert(childUsesTerminal == 0 || childUsesTerminal == 1);
+
+    if (pthread_mutex_lock(&g_lock) == 0)
+    {
+        // If the application is performing a read, assume the child process won't use the terminal.
```

Comment:
Why is this ok to assume?

Reply:
We can't be sure. By default, ProcessStartInfo gives the child access to the terminal. There should be only one program actively using the terminal at any point, and because the application itself is reading, we assume it isn't the child.

Reply:
> and because the application itself is reading, we assume it isn't the child.

What happens if we assume incorrectly?